File: modules/api/internal/core/app/greeting_service.go
package application

import (
	"errors"

	"rbi/api/internal/core/domain"
	"rbi/api/internal/port/outbound"
)

// GreetingService orchestrates the greeting use case.
type GreetingService struct {
	repository outbound.GreetingRepository
}

func NewGreetingService(repo outbound.GreetingRepository) *GreetingService {
	return &GreetingService{repository: repo}
}

func (s *GreetingService) Greet(name string) (domain.Greeting, error) {
	if name == "" {
		return domain.Greeting{}, errors.New("name is required")
	}

	greeting, err := s.repository.FindGreeting(name)
	if err != nil {
		return domain.Greeting{}, err
	}

	return greeting, nil
}

File: modules/rust_api/src/core/app/greeting_service.rs
use crate::core::domain::Greeting;
use crate::port::{inbound::GreetingUseCase, outbound::GreetingRepository};

/// GreetingService orchestrates the greeting use case.
pub struct GreetingService<R> {
    repository: R,
}

impl<R> GreetingService<R>
where
    R: GreetingRepository,
{
    pub fn new(repository: R) -> Self {
        Self { repository }
    }
}

impl<R> GreetingUseCase for GreetingService<R>
where
    R: GreetingRepository,
{
    fn greet(&self, name: &str) -> Result<Greeting, String> {
        if name.is_empty() {
            return Err("name is required".into());
        }

        self.repository.find_greeting(name)
    }
}
